### PR TITLE
[CL Incentives] Implement `setIncentiveRecords` and `getIncentiveRecords`

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Closes: #4274

## What is the purpose of the change

This PR implements a more portable version of state reads/writes relating to incentive records with the primary goal of making it simpler to abstract this logic into a separate module in the future. Specifically, it changes incentive records to work based off of pool Ids instead of as a field under pool objects.

## Brief Changelog

- Implement set/get functions for incentive records that do not depend on pool objects
- Remove the `PoolIncentives` field from CL pools

## Testing and Verifying

- The new functions/logic are tested in `incentives_test.go`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

File: proto/osmosis/concentrated-liquidity/incentive_record.proto
Comment (p0mvn) [inline proto/osmosis/concentrated-liquidity/incentive_record.proto:13]: Why do we have `poolId` as part of the record if the key in state includes pool id as well?

File: proto/osmosis/concentrated-liquidity/incentive_record.proto
Comment (p0mvn) [inline proto/osmosis/concentrated-liquidity/incentive_record.proto:17]: Same with denom

File: proto/osmosis/concentrated-liquidity/incentive_record.proto
Comment (p0mvn) [inline proto/osmosis/concentrated-liquidity/incentive_record.proto:43]: and min uptime

File: proto/osmosis/concentrated-liquidity/incentive_record.proto
Comment (p0mvn) [reply proto/osmosis/concentrated-liquidity/incentive_record.proto:13]: We should be able to avoid duplicating state in both key and the value. It should be possible to extract these fields from the key by using:
https://github.com/osmosis-labs/osmosis/blob/316005aca4dcc3c01ca8249934b9907014ff55e8/osmoutils/store_helper.go#L110

Comment (AlpinYukseloglu) [reply proto/osmosis/concentrated-liquidity/incentive_record.proto:13]: Discussed offline and came to a conclusion on a good approach tracked here: #4284

File: x/concentrated-liquidity/incentives.go
Comment (p0mvn) [inline x/concentrated-liquidity/incentives.go:245]: Does it make sense to use `MustGet` for simplicity because we aren't using the `found` flag for anything other than doing the error check?

I think `Get` and the `found` flag are useful when we need to use the flag for controlling the future execution flow. On the other hand, if we simply return an error when `!found`, we might as well use the `MustGet` instead that encapsulates the error check and panics

File: x/concentrated-liquidity/store.go
Comment (p0mvn) [inline x/concentrated-liquidity/store.go:140]: non-blocking Q: what do you think about grabbing these components from the end?

I did something similar here: https://github.com/osmosis-labs/osmosis/blob/792375a8ae5cf1b8d5ef41ac3f27bb1394abc5f4/x/concentrated-liquidity/store.go#L59-L64

However, I'm afraid that is not very extensible because any time we want to add a new proto field, this logic will break.

We might want to refactor this later to grab these components relative to the start as opposed to the end. Let me know what you think

**Review by p0mvn — APPROVED** [review]
LGTM

File: x/concentrated-liquidity/incentives.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives.go:225]: What do you mean by this comment?

File: x/concentrated-liquidity/incentives_test.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:646]: nit: can we check the output of this is nil as well

File: x/concentrated-liquidity/incentives_test.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives_test.go:615]: This test was super clear and easy to follow, nice job :)

File: x/concentrated-liquidity/store.go
Comment (czarcas7ic) [inline x/concentrated-liquidity/store.go:115]: shouldn't we catch the error here to ensure if an error occurs that the incentive record that gets returned is empty? (I assume this is done by Unmarshal itself but seems just nice to guarantee on our end)

**Review by czarcas7ic — APPROVED** [review]
👍

File: x/concentrated-liquidity/incentives.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives.go:245]: Hmm I opted for `Get` because I wanted to avoid putting a panic in a getter (this also seems to be a design pattern we follow elsewhere) – did you mean using `MustGet` and then catching the panic to return an error?

File: x/concentrated-liquidity/store.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/store.go:115]: Unmarshal does handle this, but I agree that for readability it's better to catch the error. Added!

File: x/concentrated-liquidity/incentives_test.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives_test.go:646]: Done!

File: x/concentrated-liquidity/incentives.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives.go:225]: This was from before I removed `incentiveRecord` as a pool field – deleted this comment now.

File: x/concentrated-liquidity/store.go
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/store.go:140]: I figure even if we add a proto field it is unlikely we will add it to the key (in fact if it's in the key we probably won't make it a proto field), so I don't foresee us breaking these unless we do a major refactor to pool/position protos.

This is since I'm assuming these fields will always be sufficient to identify a position (or incentive record in my case), so we will likely leave the key unchanged. Does this align with your thinking or did you mean something else?

File: x/concentrated-liquidity/incentives.go
Comment (p0mvn) [reply x/concentrated-liquidity/incentives.go:244]: I see - this sgtm. 

I was coming from the perspective of reducing the code boilerplate. However, your explanation justified keeping it as is

File: x/concentrated-liquidity/store.go
Comment (p0mvn) [reply x/concentrated-liquidity/store.go:144]: Yeah, that makes sense - agreed

**Review by p0mvn — APPROVED** [review]
Nice work. Merging since 2 approvals + all comments have been addressed

**Comment by AlpinYukseloglu** [comment]
Reverting since the base branch is already quite large & difficult to review. Made new PR #4295 that we should hold off on merged until #4231 is reviewed and merged

**Pull request merged**
